### PR TITLE
Further cleaning of ResumeManager API

### PR DESCRIPTION
**Pull request by somasun**
This PR has the following changes.
- `InMemResumeManager` has been divided into two classes - `WarmResumeManager` and `ColdResumeManager`.  `WarmResumeManager` will be the default ResumeManager used in RSocket.  `ColdResumeManager` is an in-memory implementation of ResumeManger for cold-resumption (for testing and prototyping of cold-resumption).  It extends `WarmResumeManager` and implements additional methods relevant for cold resumption.
- The cold resumption specific API has been simplified to involve only one call - `getStreamInfo()` to get all stream related information during cold-resumption.

I guess with this PR, all changes needed for basic cold-resumption of client-originated REQUEST_STREAMs are in. 

Notes:
- I will follow up by writing a disk-based ColdResumeManager
- The cmake targets needs to be simplified so that common targets can be reused by both tests and examples.

### CODE REVIEW

File: rsocket/ResumeManager.h [hunk 1]
```diff
@@ -15,9 +15,17 @@ class IOBuf;
 
 namespace rsocket {
 
+struct StreamResumeInfo {
+  StreamType streamType;
+  Requester requester;
+  std::string streamToken{""};
```
Comment (alexmalyshev) [inline rsocket/ResumeManager.h]: Nit: Don't need the `{""}` bit, default ctor of `std::string` is an empty string.

File: rsocket/internal/Common.h [hunk 2]
```diff
@@ -59,6 +59,11 @@ enum class StreamType {
   FNF,
 };
 
+enum class Requester {
```
Comment (alexmalyshev) [inline rsocket/internal/Common.h]: I can see this getting conflated with `RSocketRequester`...  Is there a different name we can use?

File: rsocket/internal/WarmResumeManager.h [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+    return impliedPosition_;
+  }
+
+  void onStreamClosed(StreamId) override{
+      // No action to perform for WarmResumeManager
+  };
+
+  const std::unordered_map<StreamId, StreamResumeInfo>& getStreamResumeInfo()
+      override {
+    LOG(FATAL) << "Not Implemented for Warm Resumption";
```
Comment (alexmalyshev) [inline rsocket/internal/WarmResumeManager.h]: Should this crash?  It seems like we should have softer failures for someone trying to do cold resumption with `WarmResumeManager`.

Comment (somasun) [reply rsocket/internal/WarmResumeManager.h]: I think crash is better here.  Users should not be attempting cold resumption with `WarmResumeManager`.   This is mostly during development.  It is unlikely that a user will suddenly hit this code path and FATAL in production.

File: rsocket/ResumeManager.h [hunk 4]
```diff
@@ -15,9 +15,42 @@ class IOBuf;
 
 namespace rsocket {
 
+// Struct to hold information relevant per stream.
+struct StreamResumeInfo {
+  StreamResumeInfo() = delete;
+  StreamResumeInfo(
+      StreamType _streamType,
```
Comment (alexmalyshev) [inline rsocket/ResumeManager.h]: Nit: Don't name variables `_foo`, all underscore-prefixed identifiers are "reserved by the system" in C++.

File: rsocket/framing/Frame.cpp [hunk 5]
```diff
@@ -249,4 +249,25 @@ std::ostream& operator<<(std::ostream& os, const Frame_ERROR& frame) {
...
+    if (frameType == FrameType::REQUEST_STREAM) {
+      return StreamType::STREAM;
+    } else if (frameType == FrameType::REQUEST_CHANNEL) {
+      return StreamType::CHANNEL;
+    } else if (frameType == FrameType::REQUEST_RESPONSE) {
+      return StreamType::REQUEST_RESPONSE;
+    } else if (frameType == FrameType::REQUEST_FNF) {
+      return StreamType::FNF;
+    } else {
+      CHECK(false) << "Unknown open stream frame : " << frameType;
```
Comment (alexmalyshev) [inline rsocket/framing/Frame.cpp]: Nit: Use `LOG(FATAL)` instead.  It's marked as `[[noreturn]]` so there won't be any compilation errors.

File: rsocket/statemachine/RSocketStateMachine.cpp [hunk 6]
```diff
@@ -461,28 +461,27 @@ void RSocketStateMachine::handleConnectionFrame(
...
-                  std::move(subscriber),
-                  *folly::EventBaseManager::get()->getEventBase()),
-              streamId,
-              consumerAllowance);
-        }
-
         streamsFactory().setNextStreamId(
             resumeManager_->getLargestUsedStreamId());
-
+        for (auto it : resumeManager_->getStreamResumeInfos()) {
```
Comment (alexmalyshev) [inline rsocket/statemachine/RSocketStateMachine.cpp]: You want `auto&`, `it` here is a `std::pair<StreamId, StreamResumeInfo>` which you don't want to copy.

File: rsocket/statemachine/RSocketStateMachine.cpp [hunk 7]
```diff
@@ -620,74 +619,67 @@ void RSocketStateMachine::handleUnknownStream(
...
-      auto stateMachine =
-          streamsFactory_.createChannelResponder(frame.requestN_, streamId);
-      auto requestSink = requestResponder_->handleRequestChannelCore(
-          std::move(frame.payload_), streamId, stateMachine);
-      stateMachine->subscribe(requestSink);
-      break;
+  if (!isNewStreamFrame(frameType)) {
+    std::stringstream msg;
+    msg << "Unexpected frame " << frameType << " for stream " << streamId;
+    LOG(ERROR) << msg.str();
```
Comment (alexmalyshev) [inline rsocket/statemachine/RSocketStateMachine.cpp]: I think this should be a `VLOG` or deleted.  I can see this getting spammy with rogue clients, and it's better to just stay away from that.

File: test/test_utils/ColdResumeManager.cpp [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+
+void ColdResumeManager::trackSentFrame(
+    const folly::IOBuf& serializedFrame,
+    FrameType frameType,
+    folly::Optional<StreamId> streamIdPtr,
+    size_t consumerAllowance) {
+  if (!shouldTrackFrame(frameType)) {
+    return;
+  }
+  const StreamId streamId = *streamIdPtr;
```
Comment (alexmalyshev) [inline test/test_utils/ColdResumeManager.cpp]: Why don't we need to check if `streamIdPtr` is `folly::none`?

Comment (somasun) [reply test/test_utils/ColdResumeManager.cpp]: All frames should have StreamID.  So I want to get rid of the `folly::Optional` part for the StreamID in another PR.  For now, will add the check.

**Pull request merged**